File: examples/dev/ir-spec/mod03.ml
type unit0'0 = unit0

and unit0 = Unit0'Unit0

and 'tA boxed'1 = 'tA boxed

and 'tA boxed = Boxed'Boxed of 'tA

and ('tA, 'tB) either'2 = ('tA, 'tB) either

and ('tA, 'tB) either = Either'Left of 'tA | Either'Right of 'tB

and rgb'0 = rgb

and rgb = Rgb'R | Rgb'G | Rgb'B

and ('tA, 'tB) pair'2 = ('tA, 'tB) pair

and ('tA, 'tB) pair = Pair'Pair of 'tA * 'tB

and ('tA, 'tB, 'tC) triple'3 = ('tA, 'tB, 'tC) triple

and ('tA, 'tB, 'tC) triple = Triple'Triple of 'tA * 'tB * 'tC

and 'tA my_list'1 = 'tA my_list

and 'tA my_list = My_list'Cons of 'tA * 'tA my_list'1 | My_list'Nil

and 'tA option'1 = 'tA option

and 'tA option = Option'None | Option'Some of 'tA

let rec mk_rgb'0 : unit -> rgb'0 = function () -> Rgb'R

let rec mk_triple'1 : 'tA -> ('tA, 'tA, 'tA) triple'3 = function
  | v_A -> Triple'Triple (v_A, v_A, v_A)

let rec mk_none'0 : unit -> _ option'1 = function () -> Option'None

let rec mk_unit'0 : unit -> unit0'0 = function () -> Unit0'Unit0

let rec mk_box'1 : 'tA -> 'tA boxed'1 = function v_A -> Boxed'Boxed v_A

let rec mk_left'1 : 'tA -> ('tA, _) either'2 = function v_A -> Either'Left v_A

let rec mk_right'1 : 'tB -> (_, 'tB) either'2 = function
  | v_B -> Either'Right v_B

let rec zero'2 : unit0'0 * 'tV -> 'tV = function Unit0'Unit0, v_Val -> v_Val

let rec unbox'1 : 'tE boxed'1 -> 'tE = function
  | v_Boxed -> ( match v_Boxed with Boxed'Boxed v_Elem -> v_Elem )

let rec un_either'1 : ('tA, 'tA) either'2 -> 'tA = function
  | v_Either -> (
      match v_Either with
      | Either'Left v_Elem -> v_Elem
      | Either'Right v_Elem -> v_Elem )

let rec un_pair'1 : ('tA, 'tB) pair'2 -> 'tA * 'tB = function
  | v_Pair -> ( match v_Pair with Pair'Pair (v_A, v_B) -> (v_A, v_B) )

let rec first'1 : ('tA, _) pair'2 -> 'tA = function Pair'Pair (v_F, _) -> v_F

let rec second'1 : (_, 'tB) pair'2 -> 'tB = function Pair'Pair (_, v_S) -> v_S

File: examples/dev/ir-spec/mod03.erl
-file("dev/src/mod03.erlt", 1).

-module(mod03).

-export_type([unit0/0,
              boxed/1,
              either/2,
              rgb/0,
              pair/2,
              triple/3,
              my_list/1,
              option/1]).

-export([mk_rgb/0,
         mk_triple/1,
         mk_none/0,
         mk_unit/0,
         mk_box/1,
         mk_left/1,
         mk_right/1,
         zero/2,
         unbox/1,
         un_either/1,
         un_pair/1,
         first/1,
         second/1,
         call_this_mod/1]).

-type unit0() :: {969696, mod03, unit0, unit0}.

-type boxed(A) :: {969696, mod03, boxed, boxed, A}.

-type either(A, B) :: {969696, mod03, either, left, A} |
                      {969696, mod03, either, right, B}.

-type rgb() :: {969696, mod03, rgb, r} |
               {969696, mod03, rgb, g} |
               {969696, mod03, rgb, b}.

-type pair(A, B) :: {969696, mod03, pair, pair, A, B}.

-type triple(A, B, C) :: {969696,
                          mod03,
                          triple,
                          triple,
                          A,
                          B,
                          C}.

-type my_list(A) :: {969696,
                     mod03,
                     my_list,
                     cons,
                     A,
                     my_list(A)} |
                    {969696, mod03, my_list, nil}.

-type option(A) :: {969696, mod03, option, none} |
                   {969696, mod03, option, some, A}.

-spec mk_rgb() -> rgb().

mk_rgb() -> {969696, mod03, rgb, r}.

-spec mk_triple(A) -> triple(A, A, A).

mk_triple(A) ->
    {969696, mod03, triple, triple, A, A, A}.

-spec mk_none() -> option(_).

mk_none() -> {969696, mod03, option, none}.

-spec mk_unit() -> unit0().

mk_unit() -> {969696, mod03, unit0, unit0}.

-spec mk_box(A) -> boxed(A).

mk_box(A) -> {969696, mod03, boxed, boxed, A}.

-spec mk_left(A) -> either(A, _).

mk_left(A) -> {969696, mod03, either, left, A}.

-spec mk_right(B) -> either(_, B).

mk_right(B) -> {969696, mod03, either, right, B}.

-spec zero(unit0(), V) -> V.

zero({969696, mod03, unit0, unit0}, Val) -> Val.

-spec unbox(boxed(E)) -> E.

unbox(Boxed) ->
    case Boxed of
        {969696, mod03, boxed, boxed, Elem} -> Elem
    end.

-spec un_either(either(A, A)) -> A.

un_either(Either) ->
    case Either of
        {969696, mod03, either, left, Elem} -> Elem;
        {969696, mod03, either, right, Elem} -> Elem
    end.

-spec un_pair(pair(A, B)) -> {A, B}.

un_pair(Pair) ->
    case Pair of
        {969696, mod03, pair, pair, A, B} -> {A, B}
    end.

-spec first(pair(A, _)) -> A.

first({969696, mod03, pair, pair, F, _}) -> F.

-spec second(pair(_, B)) -> B.

second({969696, mod03, pair, pair, _, S}) -> S.

-spec call_this_mod(pair(A, B)) -> pair(A, B).

call_this_mod(P) ->
    case P of
        {969696, mod03, pair, pair, A, B} ->
            {969696, mod03, pair, pair, A, B}
    end.


